<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Data-Consistency-Checks" data-toc-modified-id="Data-Consistency-Checks-1">Data Consistency Checks</a></span><ul class="toc-item"><li><span><a href="#01-Import-libraries" data-toc-modified-id="01-Import-libraries-1.1">01 Import libraries</a></span></li><li><span><a href="#02-Create-path" data-toc-modified-id="02-Create-path-1.2">02 Create path</a></span></li><li><span><a href="#03-import-data" data-toc-modified-id="03-import-data-1.3">03 import data</a></span></li><li><span><a href="#04-Mixed-Type-Data-Practice" data-toc-modified-id="04-Mixed-Type-Data-Practice-1.4">04 Mixed-Type Data Practice</a></span></li><li><span><a href="#05-Missing-Values" data-toc-modified-id="05-Missing-Values-1.5">05 Missing Values</a></span></li><li><span><a href="#06-Duplicates" data-toc-modified-id="06-Duplicates-1.6">06 Duplicates</a></span></li><li><span><a href="#07-Export-checked-data" data-toc-modified-id="07-Export-checked-data-1.7">07 Export checked data</a></span></li></ul></li><li><span><a href="#TASK-4.5-Data-Consistency-Checks" data-toc-modified-id="TASK-4.5-Data-Consistency-Checks-2">TASK 4.5 Data Consistency Checks</a></span><ul class="toc-item"><li><span><a href="#01.-If-you-haven’t-performed-the-consistency-checks-covered-in-this-Exercise-on-your-df_prods-dataframe,-do-so-now." data-toc-modified-id="01.-If-you-haven’t-performed-the-consistency-checks-covered-in-this-Exercise-on-your-df_prods-dataframe,-do-so-now.-2.1">01. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.</a></span></li><li><span><a href="#02-Run-the-df.describe()-function-on-your-df_prods-dataframe." data-toc-modified-id="02-Run-the-df.describe()-function-on-your-df_prods-dataframe.-2.2">02 Run the df.describe() function on your df_prods dataframe.</a></span></li><li><span><a href="#03-Check-for-mixed-type-data-in-your-df_ords-dataframe." data-toc-modified-id="03-Check-for-mixed-type-data-in-your-df_ords-dataframe.-2.3">03 Check for mixed-type data in your df_ords dataframe.</a></span></li><li><span><a href="#04-If-you-find-mixed-type-data,-fix-it.-The-column-in-question-should-contain-observations-of-a-single-data-type." data-toc-modified-id="04-If-you-find-mixed-type-data,-fix-it.-The-column-in-question-should-contain-observations-of-a-single-data-type.-2.4">04 If you find mixed-type data, fix it. The column in question should contain observations of a single data type.</a></span></li><li><span><a href="#05-Run-a-check-for-missing-values-in-your-df_ords-dataframe." data-toc-modified-id="05-Run-a-check-for-missing-values-in-your-df_ords-dataframe.-2.5">05 Run a check for missing values in your df_ords dataframe.</a></span></li><li><span><a href="#06-Address-the-missing-values-using-an-appropriate-method." data-toc-modified-id="06-Address-the-missing-values-using-an-appropriate-method.-2.6">06 Address the missing values using an appropriate method.</a></span></li><li><span><a href="#07-Run-a-check-for-duplicate-values-in-your-df_ords-data." data-toc-modified-id="07-Run-a-check-for-duplicate-values-in-your-df_ords-data.-2.7">07 Run a check for duplicate values in your df_ords data.</a></span></li><li><span><a href="#08-Address-the-duplicates-using-an-appropriate-method." data-toc-modified-id="08-Address-the-duplicates-using-an-appropriate-method.-2.8">08 Address the duplicates using an appropriate method.</a></span></li><li><span><a href="#09-Export-cleaned-dataframes-as-csv" data-toc-modified-id="09-Export-cleaned-dataframes-as-csv-2.9">09 Export cleaned dataframes as csv</a></span></li></ul></li></ul></div>

# Data Consistency Checks

## 01 Import libraries

In [58]:
# import libraries
import pandas as pd
import numpy as np
import os

## 02 Create path

In [59]:
path = r'C:\Users\rober_fjs0dbr\OneDrive\Documents\CF Coursework\Python Fundamentals for Data Analysts\11-04-2022 Instacart Basket Analysis'

In [60]:
path

'C:\\Users\\rober_fjs0dbr\\OneDrive\\Documents\\CF Coursework\\Python Fundamentals for Data Analysts\\11-04-2022 Instacart Basket Analysis'

## 03 import data

In [61]:
# import products.csv data as df_prods from orginal data folder
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [62]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [63]:
# import orders_wrangled.csv data as df_ords from prepared data folder
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [64]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

## 04 Mixed-Type Data Practice

In [65]:
# create test dataframe

df_test = pd.DataFrame()

In [66]:
# create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [67]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [68]:
# custom function which searches dataframe for mixed columns

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [69]:
# convert'mix' column from df_test to string

df_test['mix'] = df_test['mix'].astype('str')

## 05 Missing Values

#01 Finding missing values

In [70]:
# finds missing values in df_prods

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [71]:
# create subset of missing values from df_prods

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [72]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

#02 Addressing Missing Values

In [73]:
df_prods.shape

(49693, 5)

In [74]:
# create new dataframe df_prods_clean without missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [75]:
df_prods_clean.shape

(49677, 5)

## 06 Duplicates

#01 Finding duplicates

In [76]:
# command creating subset of full duplicates from df_prods_clean

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [77]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

#02 Addressing duplicates

In [78]:
#current number of rows in df_prods_clean

df_prods_clean.shape

(49677, 5)

In [79]:
# create subset without duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [80]:
df_prods_clean_no_dups.shape

(49672, 5)

## 07 Export checked data

In [81]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# TASK 4.5 Data Consistency Checks

## 01. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

See above sections

## 02 Run the df.describe() function on your df_prods dataframe. 
Using your new knowledge about how to interpret the output of this 
function, share in a markdown cell whether anything about the data 
looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [ ]:
# create output describing data in df_prods

df_prods_clean_no_dups.describe()

In [ ]:
# product_id, aisle_id, and department_id are being treated as numbers(int)
and should be treated as string values. Also, the max price of $99,999 
appears to be an imput error. I belive it should be replaced with the median
value of all prices.

In [ ]:
# set df_prods as df_prods_clean_no_dups

df_prods = df_prods_clean_no_dups

In [ ]:
df_prods.describe()

## 03 Check for mixed-type data in your df_ords dataframe.

In [ ]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)



## 04 If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

no mixed-type data exists in df_ords dataframe

## 05 Run a check for missing values in your df_ords dataframe.

In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [ ]:
# finds missing values in df_prods

df_ords.isnull().sum()

There are 206,209 missing values in the days_since_prior_order column of df_ords.
Thi

In [ ]:
# create subset of missing values from df_prods

df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [ ]:
df_ords_nan

## 06 Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

I believe these values are first purchases made by the customer so there would be no prior purchase to
relate to. I would suggest leaving this data because it may bring value to the project.

## 07 Run a check for duplicate values in your df_ords data.

In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [ ]:
# command creating subset of full duplicates from df_ords

df_dups_ords = df_ords[df_ords.duplicated()]

In [ ]:
df_dups_ords

## 08 Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

No duplicate values were found in df_ords.

## 09 Export cleaned dataframes as csv

In [ ]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [ ]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))